ringbuffer: implement efficient and performant ringbuffer

brauner opened the pull request:
liblxc will use a ringbuffer implementation that employs mmap()ed memory.
Specifically, the ringbuffer will create an anonymous memory mapping twice the
requested size for the ringbuffer. Afterwards, an in-memory file the requested
size for the ringbuffer will be created. This in-memory file will then be
memory mapped twice into the previously established anonymous memory mapping
thereby effectively splitting the anoymous memory mapping in two halves of
equal size.  This will allow the ringbuffer to get rid of any complex boundary
and wrap-around calculation logic. Since the underlying physical memory is the
same in both halves of the memory mapping only a single memcpy() call for both
reads and writes from and to the ringbuffer is needed.

Design Notes:
- Since we're using MAP_FIXED memory mappings to map the same in-memory file
  twice into the anonymous memory mapping the kernel requires us to always
  operate on properly aligned pages. To guarantee proper page aligment the size
  of the ringbuffer must always be a muliple of the kernel's page size. This
  also implies that the minimum size of the ringbuffer must be at least equal to
  one page size. This additional requirement is reasonably unproblematic.
  First, any ringbuffer smaller than the size of a single page is very likely
  useless since the standard page size on linux is 4096 bytes.
- Because liblxc is not able to predict the output a user is going to produce
  (e.g. users could cat binary files onto the console) and because the
  ringbuffer is located in a hotpath and needs to be as performant as possible
  liblxc will not parse the buffer.

Use Case:
The ringbuffer is needed by liblxc in order to safely log the output of write
intensive callers that produce unpredictable output or unpredictable amounts of
output. The console output created by a booting system and the user is one of
those cases. Allowing a container to log the console's output to a file it
would be possible for a malicious user to fill up the host filesystem by
producing random ouput on the container's console if quota support is either
not enabled or not available for the underlying filesystem. Using a ringbuffer
is a reliable and secure way to ensure a fixed-size log.

Closes #1857.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

brauner commented:
@hallyn, this only adds the internal plumbing for the ringbuffer but doesn't implement the API. This is on purpose. We should come up with two nice + extensible API functions to retrieve the current console ringbuffer and clear the current console ringbuffer.

brauner commented:
Also the name of the config key is currently `lxc.console.logsize'. Better naming suggestions very welcome!

brauner commented on src/lxc/ringbuf.c:
```diff
@@ -0,0 +1,150 @@
+		return -EINVAL;
+
+	/* For kernels < 3.17 we can fallback to mkstemp() + unlink(). Not
+	 * required for the RFC.
```
Old comment.-

stgraber commented on src/lxc/utils.c:
```diff
@@ -2339,3 +2360,132 @@ bool lxc_nic_exists(char *nic)
+{
+	int fd, ret;
+
+	fd = mkstemps(template, suffixlen);
```
jenkins doesn't like that on Android

brauner replied:
/me rolls eyes at Android.

hallyn commented on src/lxc/utils.c:
```diff
@@ -2339,3 +2339,19 @@ bool lxc_nic_exists(char *nic)
+{
+	int oflags;
+
+	oflags = fcntl(fd, F_GETFD, 0);
```
I'm probably overthinking this but...   since liblxc is sometimes used in very highly threaded contexts - is this too racy?  (By the time we get around to our F_SETFD, someone else may have changed the base flags)

I'm probably overthinking this.

brauner replied:
That's actually a good point. We might simply set the flag and call it a day...

brauner replied:
addressed

hallyn commented on src/lxc/utils.c:
```diff
@@ -2355,3 +2355,31 @@ int lxc_cloexec(int fd, bool set)
+	return fd;
+
+on_error:
+	close(fd);
```
Do you want to rm it here too?

brauner replied:
Yup.

brauner replied:
addressed

hallyn commented on src/lxc/confile.c:
```diff
@@ -1790,6 +1792,45 @@ static int set_config_console_logfile(const char *key, const char *value,
+	/* Must be at least a page size */
+	pgsz = getpagesize();
+	if (size < pgsz) {
+		ERROR("%s - The size of the console log must be at least %d "
```
Would it be better to just automatically set it to the min?

Otherwise users could get annoying start failures when moving a container to a different system.

brauner replied:
But warn users in the log?

brauner replied:
addressed

hallyn commented on src/lxc/start.c:
```diff
@@ -937,7 +937,7 @@ static int do_start(void *data)
 
 	/* Setup the container, ip, names, utsname, ... */
-	ret = lxc_setup(handler);
+	ret = lxc_setup_child(handler);
```
This belongs in the previous patch, right?

brauner replied:
addressed

hallyn commented on src/lxc/console.c:
```diff
@@ -196,6 +196,15 @@ static int lxc_console_cb_con(int fd, uint32_t events, void *data,
 		if (console->peer >= 0)
 			w = lxc_write_nointr(console->peer, buf, r);
+
+		if (console->log_size > 0) {
```
I *think* this is right, but just to make sure - is there any case where a console reader or writer will be in a different thread?

brauner replied:
@hallyn, I don't think so. We need to go through the container's command socket to request an operation on the console ringbuffer to be performed. The command socket handler and console handler will always be in the same thread since they are setup together on container in start in the epoll loop.

brauner replied:
My point being that the lock needed for an operation on the ringbuffer will always be taken by the monitor.

hallyn commented on src/lxc/conf.c:
```diff
@@ -3079,6 +3082,35 @@ static bool verify_start_hooks(struct lxc_conf *conf)
+	if (size <= 0)
+		return 0;
+
+	ret = lxc_ringbuf_create(buf, size);
```
Doesn't lxc_console_free() need to be updated?

hallyn replied:
I guess it's just an mmap so no big deal

brauner replied:
That is an interesting question. I think not because `lxc_console_free()` is only ever called when you allocate the console or ttys from an already running container. In this case the console ringbuffer will have already been setup by the parent so when you detach from the console again but you'd free the ringbuffer you'd make the parent crash since the console epoll handler will still try to write to it.

hallyn commented on src/lxc/ringbuf.c:
```diff
@@ -0,0 +1,146 @@
+
+	if (buf->r_off < buf->size)
+		return;
+	fprintf(stderr, "BLA\n");
```
?

brauner replied:
lol

brauner replied:
addressed

hallyn commented:
Thanks, looks good overall.  I think the name for lxc.console.logsize is good.

Please ping me when you feel feedback is addressed.

brauner commented:
@hallyn, updated. The changes we discussed on irc but that haven't been discussed in a comment here are:

- `parse_byte_size_string()` will consider `""` and `NULL`  invalid input and return `-EINVAL`. The redundant check for `end == dup` has been removed.
- `lxc_find_next_power2()` will consider `0` invalid input. Since it returns a `uint64_t` I have it return `0` in response to `0` as input since `0` is not a power of `2`.

brauner commented:
jenkins: test this please

brauner commented:
@hallyn, what I wonder about is where the deallocation for the console ringbuffer will need to happen. Should it happen in `lxc_conf_free()` - which means that the deallocation happens on `lxc_container_put()`- or on container stop. I'm tempted to say the deallocation needs to happen on container stop and not on `lxc_container_put()`. Because it's perfectly valid to not call `lxc_container_put()` and simply start the container again which would mean that you'd create a new ringbuffer and loose track of the old mapping without deallocation it first. Dellocating it on stop should also be fine from a users perspective because when the container has stopped the command socket handler will be gone which means - in the implementation I'm thinking about - that users cannot send any requests to get or clear the console ringbuffer anymore. They are basically blocked from accessing the console ringbuffer.
Wait, there might be a smarter option... Currently the console ringbuffer is initialized on start when the check `if (console->log_size > 0) allocate_console_ringbuffer()` is true. We should simply check whether the ringbuffer already has a valid address aka `if (console->ringbuffer.addr) resuse_buffer()`. Yeah, that's the right thing to do. But then the question becomes do we clear the buffer on startup or do we keep the former read and write offset...
I'll wait for you to comment before I make any changes. :) But I think the second option (reusing the buffer) is correct. So deallocation can happen on `lxc_container_put()`...

brauner commented:
Hm, you need to account for the following API sequences:

- `c = lxc_container_new()`
- `c->start()` /* logsize == 4096 */
- `c->stop()`
- `c->start()` /* logsize == 4096 --> reuse ringbuffer */
- `c->stop()`
- `c->clear_config()`
- `c->load_config()`
- `c->start()` /* logsize == 8096 --> deallocate  old ringbuffer --> allocate newringbuffer */


`clear_config()` and `load_config()` where the logsize has changed. But that's fine in this case we can just deallocate the current buffer and initialize a new one...

hallyn commented:
Quoting Christian Brauner (notifications@github.com):
> brauner commented on this pull request.
> 
> 
> 
> > @@ -196,6 +196,15 @@ static int lxc_console_cb_con(int fd, uint32_t events, void *data,
>  
>  		if (console->peer >= 0)
>  			w = lxc_write_nointr(console->peer, buf, r);
> +
> +		if (console->log_size > 0) {
> 
> My point being that the lock needed for an operation on the ringbuffer will always be taken by the monitor.

so, is the process_lock() needed at all then?  The monitor will
always be the same thread?

hallyn commented:
... well, if we are re-using the buffer, the user *may* want to still see output
from the last (perhaps failed) boot?  

dealloc on put sounds good.

brauner commented:
> so, is the process_lock() needed at all then?  The monitor will
> always be the same thread?

I think you're right that we don't need the lock. But let's see if my reasoning is sounds. Both the command socket handler and the console ringbuffer handler run in the same epoll loop in the same thread. Now we send a request `console_get_log()` which will cause a handler to be register via the command socket in the container's epoll loop. Now either the console ringbuffer handler runs or the `console_get_log()` handler runs but never both so they can't race. Does that make sense?

brauner commented:
@hallyn, added the realloc-dealloc logic to `lxc_setup_console_ringbuf()` called in `lxc_setup_parent()` and removed the `process_{un}lock()` calls. If you have no further objections this should be good to go.

brauner commented on src/lxc/ringbuf.c:
```diff
@@ -0,0 +1,145 @@
+	buf->w_off = 0;
+
+	/* verify that we are at least given the multiple of a page size */
+	if (buf->size % getpagesize())
```
function `getpagesize()` is deprecated

brauner commented:
jenkins: test this please
